modules/tftp/server.rb

```diff
@@ -90,21 +107,93 @@ def pxe_default
     end
 
     def pxeconfig_file(mac)
-      ["#{pxeconfig_dir}/menu.lst.01" + mac.delete(':').upcase, "#{pxeconfig_dir}/01-" + mac.tr(':', '-').upcase]
+      ["#{pxeconfig_dir}/menu.lst.01" + mac.delete(':').upcase, "#{pxeconfig_dir}/01-" + dashed_mac(mac).upcase]
     end
   end
 
   class Pxegrub2 < Server
-    def pxeconfig_dir
-      "#{path}/grub2"
+    def bootloader_path(os, version, arch)
+      unless (bootloader_universe = Proxy::TFTP::Plugin.settings.bootloader_universe)
+        logger.debug "TFTP: bootloader universe not configured."
+        return
+      end
+
+      bootloader_path = File.join(bootloader_universe, 'pxegrub2', os, version, arch)
+
+      logger.debug "TFTP: Checking bootloader universe for suitable bootloader directory for"
+      logger.debug "TFTP:   * Operating system: #{os}"
+      logger.debug "TFTP:   * Version: #{version}"
+      logger.debug "TFTP:   * Architecture: #{arch}"
+      logger.debug "TFTP: Checking bootloader universe if \"#{bootloader_path}\" exists."
+      unless Dir.exist?(bootloader_path)
+        logger.debug "TFTP: Directory \"#{bootloader_path}\" does not exist."
+
+        bootloader_path = "#{bootloader_universe}/pxegrub2/#{os}/default/#{arch}"
+        logger.debug "TFTP: Checking if fallback directory at \"#{bootloader_path}\" exists."
+        unless Dir.exist?(bootloader_path)
```

Comment:
We should create it automatically or have a task/migration for it. But let's not block this PR on it, created a Jira task for it: https://issues.redhat.com/browse/SAT-24819

Reply:
What do you mean by creating it automatically?
The `bootloader_path` indicates if the SB feature is enabled (empty means it's disabled).
I agree that a better approach would be to have a boolean value to turn this feature on or off and use another parameter to store the value of the `bootloader_path` as we call it today.
The default value of this path can be `/var/lib/tftpboot/`, right?

Reply:
The `bootloader_path` here is the concrete directory that contains bootloader files for a specific operating system, architecture, and version. These files have to be maintained by the users. This directory is nothing we should create automatically in case it doesn't exist, as

1. it is not possible to create the `bootloader_path` for all supported OSes and versions.
2. the non-existence of this directory is part of the logic. 

Following scenario:

1. A user activates the Bootloader Universe and wants to deploy SecureBoot-enabled AlmaLinux hosts. To do so, the user has to place AlmaLinux binaries in the Bootloader Universe.
2. During provisioning the Smart Proxy checks if there is a directory for AlmaLinux in the Bootloader Universe and delivers those files to the host upon provisioning / boot.
3. The user wants to deploy Rocky Linux hosts without Secure Boot through the same Smart Proxy. There is no need for the user to provide binaries in the Bootloader Universe for Rocky Linux (as Secure Boot is not enabled).
4. During provisioning the Smart Proxy checks if there is a directory for Rocky Linux in the Bootloader Universe. As there is none, the Smart Proxy delivers the default shim and GRUB2 binaries from the `grub2` directory inside the TFTP root.

=> As the check in step 4 above relies on the existence of the OS specific directory in the Bootloader Universe, the empty directory would be copied to the MAC address specific directory of the Rocky Linux host and the host would not be able to boot.

Small addition to what Nofar said:
The `bootloader_path` does not tell if the SB feature is enabled or not but the Smart Proxy setting of the bootloader_universe](https://github.com/theforeman/smart-proxy/pull/877/files/99a3b27003e39094aacc859b033bb155de8a16f8#diff-db3f0d8d32c77f2719de93fd56902295e1d21a1a1c0dc1e19f398cf94d69426eR116) does.
And I would also agree that this could be handled simpler, but I would even go one step further beyond what Nofar said: I would appreciate if the usage of the MAC address specific host directory would become the default way for storing bootloader files for a host (at least for the `Grub2 UEFI` PXE loaders). This would make the capability obsolete and would make the implementation simpler. I am looking forward to work on a future PR for that. :smile:

Reply:
>  I would appreciate if the usage of the MAC address specific host directory would become the default way for storing bootloader files for a host (at least for the Grub2 UEFI PXE loaders). This would make the capability obsolete and would make the implementation simpler. I am looking forward to work on a future PR for that.

That would be a great feature, and I agree doing it in follow up PRs

Reply:
> I would appreciate if the usage of the MAC address specific host directory would become the default way for storing bootloader files for a host (at least for the Grub2 UEFI PXE loaders). This would make the capability obsolete and would make the implementation simpler. I am looking forward to work on a future PR for that.

I support that too.

Still we need to take care of network setups where DHCP (thus the NBP path) is not under Foreman's control. This requires to always provide at least `/var/lib/tftpboot/grub2/grubx64.efi` (maybe also shim for compatibility reasons). What we can do is to get rid of host specific GRUB2 configuration files in `/var/lib/tftpboot/grub2/` and to only provide generic `/var/lib/tftpboot/grub2/grub.cfg` which then loads the host specific configuration files from `/var/lib/tftpboot/host_config/<MAC>/grub2/grub.cfg`.

But all of this can then be discussed in the corresponding PR making MAC address specific host directory the default.